flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -529,6 +530,13 @@ protected void processInput(MailboxDefaultAction.Controller controller) throws E
     }
 
     protected void endData() throws Exception {
+
+        if (syncSavepointWithoutDrain != null) {
```

Comment:
Perhaps we only do the check if checkpoints after tasks finished are enabled~? If not, the task would not wait for the final checkpoint and would not be blocked, thus we might not change the behavior if flip-147 is not enabled~